scipy/linalg/_decomp_svd.py

```diff
@@ -295,6 +299,99 @@ def diagsvd(s, M, N):
+) -> tuple[list[np.ndarray], np.ndarray]:
+    """Higher-order SVD (HOSVD)
+
+    Factorizes the M-D tensor `a` into a list of M matrices ``U_k`` containing
+    the left singular vectors of the unfolded tensor with respect to the
+    ``k``\\ th axis and a core tensor ``S`` such that
+    :math:`a = (U_1, U_2, \\dots, U_M) \\cdot S`.
+
+    Parameters
+    ----------
+    a : array_like
+        Tensor to decompose
+    compact : bool,  default: False
+        If False (default), ``S`` is of the same shape as the input tensor `a`
+        and all ``U_k`` are of the shape ``(n_k, n_k)``, where ``n_k = a.shape[k]``.
+        If True, ``S`` has shape ``m_1, ..., m_k, ... m_M`` and the shapes of
+        the ``U_k`` are ``(n_k, m_k)``, where ``n_k`` is as above and the ``m_k``
+        are the elements of multilinear rank of the `a`.
+    check_finite : bool, default: True
+        Whether to check that the input matrix contains only finite numbers.
+        Disabling may give a performance gain, but may result in problems
+        (crashes, non-termination) if the inputs do contain infinities or NaNs.
+
+    Returns
+    -------
+    U : list of ndarray
+        List of the unitary matrices ``U_k`` with shapes as described in
+        documentation of `compact`.
+    S : ndarray
+        The core tensor with shapes as described in the documentation of
+        `compact`.
+
+    Raises
+    ------
+    LinAlgError
+        If SVD computation does not converge.
+
+    .. versionadded:: 1.15.0
+
+    See Also
+    --------
+    svd : Singular value decomposition of a matrix
+
+    References
+    ----------
+    .. [1] Lieven De Lathauwer, Bart De Moor, and Joos Vandewalle (2000)
+           "A Multilinear Singular Value Decomposition"
+           SIAM Journal on Matrix Analysis and Applications,
+           21 (4), pp. 1253-1278. ISSN 1095-7162
+           https://doi.org/10.1137/S0895479896305696
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from scipy import linalg
+    >>> rng = np.random.default_rng()
+    >>> a = rng.random((2, 3, 4, 5))
+    >>> U, S = linalg.higher_order_svd(a)
+    >>> S.shape
+    (2, 3, 4, 5)
+    >>> [i.shape for i in U]
+    [(2, 2), (3, 3), (4, 4), (5, 5)]
+
+    Reconstruct the original tensor:
+
+    >>> a1 = S
+    >>> for i, _ in enumerate(U):
+    ...     a1 = np.tensordot(a1, U[i], (0, 1))
+    >>> np.allclose(a, a1)
+    True
+    """
+    a = _asarray_validated(a, check_finite=check_finite)
+    if compact not in {True, False}:
+        raise ValueError("`compact` must be either `True` or `False`.")
+
+    core_tensor = a
+    left_singular_basis = []
+
+    for k in range(a.ndim):
+        newshape = (a.shape[k], math.prod(a.shape[:k]) * math.prod(a.shape[k + 1:]))
+        unfold = np.reshape(np.moveaxis(a, k, 0), newshape)
+        U, s, _ = svd(unfold, check_finite=False)
+        # todo: replace this with appropriate tolerance for vanishing singular values
+        tol = 1e-12
```

Comment:
@ilayn Can you remind me of the preferred recipe for determining whether a singular value is zero?

Reply:
Out of numerical experience vanishing singular values are typical in the area of 1e-15 times the largest singular value. But I am not sure if the LAPACK methods have a more deterministic way to determine that.

Reply:
Yeah, there is precedent somewhere that I wanted to follow.